### PR TITLE
Usergrid 1079 2x

### PR DESCRIPTION
**Pull request by tnine**
Cherry picks branch from USERGRID-1079-1x into this branch, also adds PUT functionality, and tests the password GET + PUT works correctly

### CODE REVIEW

File: stack/core/src/main/java/org/apache/usergrid/persistence/CredentialsInfo.java [hunk 1]
```diff
@@ -172,4 +180,42 @@ public int compareTo( CredentialsInfo o ) {
...
+        final boolean recoverable = ( boolean ) json.get( "recoverable");
+
+        final boolean encrypted = ( boolean ) json.get("encrypted");
+        final String hashType = ( String ) json.get( "hashType");
+        final long created = ( long ) json.get( "created");
+        final String secret = ( String ) json.get(  "secret" );
+
+        final List<String>  cryptoChain = ( List<String> ) json.get( "cryptoChain");
+
+        Preconditions.checkNotNull(created, "created is required");
```
Comment (michaelarusso) [inline stack/core/src/main/java/org/apache/usergrid/persistence/CredentialsInfo.java]: do we need to check encrypted and hashType for null values as well?

File: stack/rest/src/main/java/org/apache/usergrid/rest/applications/users/UserResource.java [hunk 2]
```diff
@@ -164,6 +167,81 @@ public JSONWithPadding setUserPasswordPut( @Context UriInfo ui, Map<String, Obje
         return new JSONWithPadding( response, callback );
     }
 
+    @GET
+    @RequireSystemAccess
+    @Path("credentials")
+    public JSONWithPadding getUserCredentials(@QueryParam("callback") @DefaultValue("callback") String callback )
+            throws Exception {
+
+        logger.info( "UserResource.setUserPassword" );
```
Comment (michaelarusso) [inline stack/rest/src/main/java/org/apache/usergrid/rest/applications/users/UserResource.java]: this is the GET path

File: stack/rest/src/main/java/org/apache/usergrid/rest/applications/users/UserResource.java [hunk 3]
```diff
@@ -164,6 +167,81 @@ public JSONWithPadding setUserPasswordPut( @Context UriInfo ui, Map<String, Obje
...
+
+
+        final ApiResponse response = createApiResponse();
+        response.setAction( "get user password" );
```
Comment (michaelarusso) [inline stack/rest/src/main/java/org/apache/usergrid/rest/applications/users/UserResource.java]: maybe let's be more explicit on what this returns as it's not the password in plaintext. 'get user password hash' or something?

File: stack/rest/src/test/java/org/apache/usergrid/rest/applications/collection/users/UserResourceIT.java [hunk 4]
```diff
@@ -1103,4 +1109,107 @@ public void queryForUserUuids() throws Exception {
...
+
+
+        //get out the hash and change it so we can validate
+        final String originalSecret = ( String ) credentials.get( "secret" );
+
+
+        //here we modify the hash a little, this way we can break password validation, then re-set it to ensure we're actually updating the credentials info correctly.
+        final String borkedSecret = originalSecret.substring( 0, originalSecret.length() -1 );
+
+        credentials.put( "credentials", borkedSecret );
```
Comment (mdunker) [inline stack/rest/src/test/java/org/apache/usergrid/rest/applications/collection/users/UserResourceIT.java]: Is this right? Or borked?

File: stack/core/src/main/java/org/apache/usergrid/persistence/CredentialsInfo.java (hunk 1 above)
Comment (tnine) [inline stack/core/src/main/java/org/apache/usergrid/persistence/CredentialsInfo.java]: No, they're usually null

**Pull request merged**